### ISSUE TITLE
Replica number going beyond what is specified

### ISSUE THREAD

**Issue by drmrbrewer** [caprover/caprover#896]
This is what I see when I run `docker service ls` for a particular app of mine:

    ID                  NAME                        MODE                REPLICAS            IMAGE                                         
    eeqo0ijv2f8z        srv-captain--myapp          replicated          3/1                 registry.myserver.mydomain.com:996/captain/img-captain-myapp:47

This is odd... `3/1` instances are running?  The instance count for my app is set to `1`:

![image](https://user-images.githubusercontent.com/9009792/98274809-cde23000-1f8b-11eb-8cc4-65f7bbdcb5dd.png)

Now, when I hit `Save & Update` in the console to restart this service, I then see yet another instance appear:

    ID                  NAME                        MODE                REPLICAS            IMAGE                                         
    eeqo0ijv2f8z        srv-captain--myapp          replicated          4/1                 registry.myserver.mydomain.com:996/captain/img-captain-myapp:47

To confirm, there are indeed four instances running:

	# docker service ps srv-captain--myapp | grep Running                                                                
	wmficopmusq1        srv-captain--myapp.1       registry.myserver.mydomain.com:996/captain/img-captain-myapp:47   server-01          Running             Running 3 minutes ago                                       
	7s59yvje4rlk         \_ srv-captain--myapp.1   registry.myserver.mydomain.com:996/captain/img-captain-myapp:47   server-01          Running             Running 8 minutes ago                                       
	ibsnq0ag91xy         \_ srv-captain--myapp.1   registry.myserver.mydomain.com:996/captain/img-captain-myapp:47   server-01          Running             Running 2 days ago                                          
	rpg9ww3aowkg         \_ srv-captain--myapp.1   registry.myserver.mydomain.com:996/captain/img-captain-myapp:47   server-01          Running             Running 2 months ago  

What is going on, and how do I prevent this happening?

My CapRover version:

![image](https://user-images.githubusercontent.com/9009792/98275509-9cb62f80-1f8c-11eb-94a3-da0a603a0aee.png)

When I re-deploy this app from the command line, the number of running instances does revert to `1/1` and then when I `Save & Update` from the console it does seem to stick at `1/1` (with a newer instance being activated).  So maybe it is only when the number of instances goes over quota e.g. `2/1` that each `Save & Update` increments it forever `3/1` `4/1` etc?  Not sure how it ever got to `2/1` in the first place.  The only thing I've done recently that is remotely unusual is a disk cleanup as described [here](https://caprover.com/docs/disk-cleanup.html), including images, volumes _and_ registry... oh and updating CapRover to `1.8.1` and then soon after to `1.8.2`.

**Comment by githubsaturn** [comment]
TLDR: it's not something that CapRover interfere with at all. It's totally a Docker issue.

My "guess" is your containers cannot be shut down - either a bug in your containers, or a bug in Docker. Knowing that it only happens for a particular app, it makes me think that there is an issue with that particular container. Some process somewhere is not allowing that container to shut down. 

> The only thing I've done recently that is remotely unusual is a disk cleanup as described here, including images, volumes and registry

Disk cleanup as you mentioned isn't relevant here at all. 

> oh and updating CapRover to 1.8.1 and then soon after to 1.8.2.

Yes, 1.8.1 didn't include all fixes that it was supposed to have. So 1.8.2 was cut shortly after. 


Related Docker issues:
https://github.com/moby/moby/issues/24394
https://github.com/docker/swarmkit/issues/1372
https://github.com/moby/moby/issues/24670

A few things for you to try out:
- If you have multiple nodes, does this happen on a particular node or not? 
- Collect your Dockerfile and docker info and open an issue on [Docker's github](https://github.com/moby/moby) to see if they have more info. Make sure to include the output of some relevant commands like:
```
docker service ps --no-trunc servicename
docker service inspect servicename
```

**Comment by drmrbrewer** [comment]
> Knowing that it only happens for a particular app

It doesn't seem to be app specific.  I have several apps in this state, and the same thing happens for all of them.  I'll investigate further.

**Comment by drmrbrewer** [comment]
> Disk cleanup as you mentioned isn't relevant here at all

Not even registry cleanup?  I have a feeling that the misbehaving apps are the ones which I haven't re-deployed since wiping the registry.  And when I restarted the following app (a different one to before) from the console (to get to `3/1` replicas), I see:

	ID                          NAME                       IMAGE                                                             NODE                DESIRED STATE       CURRENT STATE            ERROR                                                                           
	tm9bkrqz5g98qgjxl45a53peq   srv-captain--myapp.1       registry.myserver.mydomain.com:996/captain/img-captain-myapp:15   mercury-01          Running             Running 4 minutes ago                                                                                                    
	lb6ohip470wdoeyynfmo9hurq    \_ srv-captain--myapp.1   registry.myserver.mydomain.com:996/captain/img-captain-myapp:15   mercury-04          Shutdown            Rejected 4 minutes ago   "No such image: registry.myserver.mydomain.com:996/captain/img-captain-myapp:15"   
	01j13442u9rn8w0swlv32kwx6    \_ srv-captain--myapp.1   registry.myserver.mydomain.com:996/captain/img-captain-myapp:15   mercury-04          Shutdown            Rejected 5 minutes ago   "No such image: registry.myserver.mydomain.com:996/captain/img-captain-myapp:15"   
	bbh7ppm5elel902sctd8euxbv    \_ srv-captain--myapp.1   registry.myserver.mydomain.com:996/captain/img-captain-myapp:15   mercury-01          Running             Running 3 days ago                                                                                                       
	guqzogfiuneetruvfxj375mz1    \_ srv-captain--myapp.1   registry.myserver.mydomain.com:996/captain/img-captain-myapp:15   mercury-01          Running             Running 2 months ago        

Not sure how is can be `Rejected` twice (reason: `No such image`) and then magically a moment later not rejected, with another instance started up?  Unless it is somehow just replicating the existing running instance, creating too many as a result.

**Comment by githubsaturn** [comment]
Ah! Yea it could be it. I didn't realize that you wiped out an in-use image.

> Not sure how is can be Rejected twice (reason: No such image) and then magically a moment later not rejected

The rejected one is on `mercury-04` whereas the one that is not rejected is on `mercury-01`

If you're running a prod heavy cluster that you have to clean up registry for every now and then. I suggest you either use a 3rd party docker image registry (gitlab is relatively cheap). Or, you use a web GUI for docker registry and only remove the old images. If you google, there is a million of them...

I've been wanting to add one as a one click app for a while, but I haven't found one that is easy to be built with a one click app template. Someone created a PR a while ago, but since then, they're MIA 
https://github.com/caprover/one-click-apps/pull/269

**Comment by drmrbrewer** [comment]
Why would wiping out an in-use image from the self-hosted registry cause the app instances to keep on growing on the master node each time I restart from CapRover console?

**Comment by githubsaturn** [comment]
Technically, it shouldn't do that. But it looks like an edge case bug on Docker. I assume Docker swarm gets confused when it can't move a service to another node. Bottom line, yes, it shouldn't happen, and Docker swarm needs to handle this edge case better.
